File: 35. Search Insert Position/python/attempt_2.py
from typing import List
import math

class Solution:
    
    def binarySearch(self, nums: List[int], start_index: int, end_index: int, target: int) -> int:
        if start_index == end_index:
            return start_index
        
        mid = (start_index + end_index) // 2
        mid_value = nums[mid]

        if target == mid_value:
            return mid
        elif target < mid_value:
            return self.binarySearch(self, nums, start_index, mid, target)
        else:
            return self.binarySearch(self, nums, mid + 1, end_index, target)

    def searchInsert(self, nums: List[int], target: int) -> int:
        return self.binarySearch(self, nums, 0, len(nums), target)

File: 704. Binary Search/typescript/attempt_2.ts
function search(nums: number[], target: number): number {
	let l = 0;
	let r = nums.length - 1;

	while (r >= l) {
		const mid = Math.floor((r + l) / 2);
		if (nums[mid] === target) return mid;
		if (nums[mid] < target) l = mid + 1
		if (nums[mid] > target) r = mid - 1
	}

	return -1
}

export default search;
